### PR TITLE
DC-24: Implement batching to reduce the number of requests to the gateway

### PR DESCRIPTION
**Pull request by patrick-schweizer**
Jira: https://issues.opennms.org/browse/DC-24

### CODE REVIEW

**Review by agalue — APPROVED** [review]
I didn't test at scale, but on the simple end-to-end environment I've been using, I can see that OpenNMS is sending one big request with lots of metrics instead of several small ones with fewer metrics, meaning the logic is working as expected.

I'd say merge it. I'll do a scaling test next week.

**Comment by agalue** [comment]
In another test environment where I have 10 OpenNMS servers pushing data, I updated the KAR with the changes from this PR on all of them. I can see a significant reduction in the number of requests against the Gateway, so batching helped, increasing performance.

File: plugin/src/main/java/org/opennms/plugins/cloud/tsaas/TsaasStorage.java [hunk 1]
```diff
@@ -168,14 +169,25 @@ private static Timestamp toTimestamp(Instant instant) {
...
-                .collect(Collectors.toList());
-        Tsaas.Samples samplesMessage = Tsaas.Samples.newBuilder()
-                .addAllSamples(mappedSamples)
-                .build();
-        LOG.trace("Storing the following samples: {}", samplesMessage);
-        clientStub.store(samplesMessage);
+                .forEach(this.queue::add);
+
+        // send batched messages while the queue is fuller than batch size
+        while(this.queue.size() >= this.config.getBatchSize()) {
```
Comment (j-white) [inline plugin/src/main/java/org/opennms/plugins/cloud/tsaas/TsaasStorage.java:179]: We should also consider adding a timer here that will flush this periodically: imagine sending 1001 samples, the 1000 will be flushed, and then the 1 will sit there... until 999 more show up.

**Comment by agalue** [comment]
With the latest changes, I can see the timeout kicking in:
{code}

**Pull request merged**